core/libp2p/connection/capable_connection.hpp

```diff
@@ -6,20 +6,53 @@
 #ifndef KAGOME_CAPABLE_CONNECTION_HPP
 #define KAGOME_CAPABLE_CONNECTION_HPP
 
-#include "libp2p/connection/raw_connection.hpp"
+#include <functional>
+
 #include "libp2p/connection/secure_connection.hpp"
-#include "libp2p/connection/stream.hpp"
 
 namespace libp2p::connection {
+  struct Stream;
 
-  // connection that provides basic libp2p requirements to the connection
+  /**
+   * Connection that provides basic libp2p requirements to the connection: it is
+   * both secured and muxed (streams can be created over that connection)
+   */
   struct CapableConnection : public SecureConnection {
+    using StreamHandler = void(outcome::result<std::shared_ptr<Stream>>);
+    using StreamHandlerFunc = std::function<StreamHandler>;
+
+    using NewStreamHandlerFunc = std::function<void(std::shared_ptr<Stream>)>;
+
     ~CapableConnection() override = default;
 
     /**
-     * @brief Opens new stream using this connection.
+     * Start to process incoming messages for this connection
+     * @note non-blocking
+     */
+    virtual void start() = 0;
+
+    /**
+     * Stop processing incoming messages for this connection without closing the
+     * connection itself
+     * @note calling 'start' after 'close' is UB
+     */
+    virtual void stop() = 0;
+
+    /**
+     * @brief Opens new stream using this connection
+     * @param cb - callback to be called, when a new stream is established or
+     * error appears
+     */
+    virtual void newStream(StreamHandlerFunc cb) = 0;
+
+    /**
+     * @brief Set a handler, which is called, when a new stream arrives from the
+     * other side
+     * @param cb, to which a received stream is passed
+     * @note if a handler is not set, all received streams will be immediately
+     * reset
      */
-    virtual outcome::result<std::shared_ptr<Stream>> newStream() = 0;
+    virtual void onStream(NewStreamHandlerFunc cb) = 0;
```

Comment:
NewStreamHandlerFunc - why it does not accept result? stream here is guaranteed to be created and valid?

Reply:
Yes, because YamuxedConnection accepts it, registers and only then giving it to the callback, it always is valid